core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -425,14 +429,28 @@ private[log] class Cleaner(val id: Int,
           }
           messagesRead += 1
         } else {
-          val messages = ByteBufferMessageSet.deepIterator(entry.message)
+          // We use absolute offset to compare decide whether retain the message or not. This is handled by
+          // deep iterator.
```

Comment:
It seems that we may need to change the message format when copying out uncompressed messages too.